[view-transitions] Support script-originated animations

nt1m opened the pull request:
#### fd1700a9494c4ead2a8d7add78378403081646b7
<pre>
[view-transitions] Support script-originated animations
<a href="https://bugs.webkit.org/show_bug.cgi?id=269468">https://bugs.webkit.org/show_bug.cgi?id=269468</a>
<a href="https://rdar.apple.com/123016183">rdar://123016183</a>

Reviewed by Matt Woodrow and Matthieu Dubet.

- Switch to use pseudoElementIdentifier in order to support view transition pseudo-elements
- Pass in a document to the relevant CSSSelectorParserContext so the view transition setting state is passed over

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/pseudo-get-computed-style-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/pseudo-get-computed-style.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/web-animation-pseudo-incorrect-name-expected.txt:
* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::pseudoElement const):
(WebCore::KeyframeEffect::setPseudoElement):
* Source/WebCore/animation/StyleOriginatedAnimationEvent.cpp:
(WebCore::StyleOriginatedAnimationEvent::StyleOriginatedAnimationEvent):
(WebCore::StyleOriginatedAnimationEvent::pseudoElement):
* Source/WebCore/animation/WebAnimationUtilities.cpp:
(WebCore::pseudoElementIdentifierAsString):
(WebCore::pseudoElementIdentifierFromString):
(WebCore::pseudoIdAsString): Deleted.
(WebCore::pseudoIdFromString): Deleted.
* Source/WebCore/animation/WebAnimationUtilities.h:

Canonical link: <a href="https://commits.webkit.org/274961@main">https://commits.webkit.org/274961@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1d21445eb037923e6f0aa3c48a4f9d118c3b91d8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/40521 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/19533 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/42899 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/43073 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/36610 "Built successfully") 
| [❌ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/42828 "Failed to checkout and rebase branch from PR 24523") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/22493 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/16863 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/33620 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [❌ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/41095 "Failed to checkout and rebase branch from PR 24523") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/16482 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/34935 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/14202 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/14279 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/35908 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/44347 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/36745 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/36236 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/39974 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/15336 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/12571 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/38288 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/16955 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/35190 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/17005 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5372 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/16599 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mdubet reviewed (approved):
The `std::pair<bool, std::optional>` still looks weird, but otherwise LGTM

mattwoodrow commented on Source/WebCore/animation/WebAnimationUtilities.cpp:
```diff
@@ -335,34 +333,29 @@ String pseudoIdAsString(PseudoId pseudoId)
 }
 
-std::optional<PseudoId> pseudoIdFromString(const String& pseudoElement)
+std::pair<bool, std::optional<Style::PseudoElementIdentifier>> pseudoElementIdentifierFromString(const String& pseudoElement, Document* document)
```
It's really unclear (to me) what the difference between the `optional` being std::nullopt, and `PseudoId::None`. Even the bool is pretty non-obvious, except when looking at callers that provide a variable name that shows it's for parse failures.

I realise that isn't new to this patch, and you're just expanding usage, but it'd be nice for a comment to briefly cover the various 'no' states.

nt1m committed:
[view-transitions] Support script-originated animations
https://bugs.webkit.org/show_bug.cgi?id=269468
rdar://123016183

Reviewed by Matt Woodrow and Matthieu Dubet.

- Switch to use pseudoElementIdentifier in order to support view transition pseudo-elements
- Pass in a document to the relevant CSSSelectorParserContext so the view transition setting state is passed over

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/pseudo-get-computed-style-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/pseudo-get-computed-style.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/web-animation-pseudo-incorrect-name-expected.txt:
* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::pseudoElement const):
(WebCore::KeyframeEffect::setPseudoElement):
* Source/WebCore/animation/StyleOriginatedAnimationEvent.cpp:
(WebCore::StyleOriginatedAnimationEvent::StyleOriginatedAnimationEvent):
(WebCore::StyleOriginatedAnimationEvent::pseudoElement):
* Source/WebCore/animation/WebAnimationUtilities.cpp:
(WebCore::pseudoElementIdentifierAsString):
(WebCore::pseudoElementIdentifierFromString):
(WebCore::pseudoIdAsString): Deleted.
(WebCore::pseudoIdFromString): Deleted.
* Source/WebCore/animation/WebAnimationUtilities.h:

Canonical link: https://commits.webkit.org/274961@main
